Learning task #1

lausiawsam opened the pull request:
Building on top of the dummy app I have so far:
- Lay out a mock screen without Neptune - using UIKit from scratch
- Add a simple view controller to demonstrate a counter that can be incremented and decremented (shown when tapping on the first navigation option)

https://github.com/lausiawsam/onboarding-app/assets/137056262/086773c1-efc1-461c-b3c0-e4a2aadba053

siawsamlau-art committed:
Add .gitignore

siawsamlau-art committed:
Layout learning screen UI.

siawsamlau-art committed:
Add counter screen.

lausiawsam commented on onboarding-app/SecondViewController.swift:
```diff
@@ -3,27 +3,178 @@ import UIKit
+        
+        var buttonConfig = UIButton.Configuration.plain()
+        buttonConfig.image = closeButtonImage
+        // TODO: ask why there's a horizontal margin around the image by default. Is this a default UIButton inset thing? Or is it an SFSymbol thing?
```
I have a question here: I noticed that there was quite a bit of horizontal space around the button image by default. However, googling tells me that there shouldn't be that space by default. So I'm wondering, is this something from SF Symbols instead?

InderKumarRathore replied:
let's discuss visually :D not getting what you are explaining here

InderKumarRathore commented on onboarding-app/CounterViewController.swift:
```diff
@@ -0,0 +1,70 @@
+import Foundation
+import UIKit
+
+class CounterViewController: UIViewController {
```
we make classes final which are not intended to be subclassed. Adding final keyword to class does magic. Find that magic and let me know.

InderKumarRathore commented on onboarding-app/CounterViewController.swift:
```diff
@@ -0,0 +1,70 @@
+
+class CounterViewController: UIViewController {
+    
+    let stackView = UIStackView()
+    let counterLabel = UILabel()
+    var counter = 0
```
could be private ?

InderKumarRathore commented on onboarding-app/CounterViewController.swift:
```diff
@@ -0,0 +1,70 @@
+    }
+    
+    @objc func decrementCounter() {
+        self.counter -= 1
```
we don't use explicit self in the code
```suggestion
        counter -= 1
```

InderKumarRathore commented on onboarding-app/SecondViewController.swift:
```diff
@@ -3,27 +3,178 @@ import UIKit
+        
+        stackView.addArrangedSubview(titleLabel)
+        NSLayoutConstraint.activate([
+            titleLabel.leadingAnchor.constraint(equalTo: stackView.layoutMarginsGuide.leadingAnchor),
+            titleLabel.trailingAnchor.constraint(equalTo: stackView.layoutMarginsGuide.trailingAnchor)
```
not sure why you need this, usually when adding to stack view we don't add constraints to child view (except if setting height/width).

lausiawsam committed:
Address PR comments; extract navigation option to a separate view

lausiawsam commented:
- Addressed the comments above
- Changed leading and trailing edge constraints to width constraints where applicable
- Extracted the navigation option out into a separate view file

InderKumarRathore commented on onboarding-app/NavigationOptionView.swift:
```diff
@@ -0,0 +1,80 @@
+        firstLineText: String,
+        secondLineText: String,
+        imageSystemName: String,
+        viewController: Any? = nil,
```
better to have a closure like `onTap` instead of vc. but before changing anything let's run the memory profile and check if there is memory leaks

lausiawsam committed:
Change to closure approach to pass tap gesture handler to NavigationOption

lausiawsam commented on onboarding-app/SecondViewController.swift:
```diff
@@ -103,13 +103,16 @@ final class SecondViewController: UIViewController {
-        onTap: Selector? = nil
+        showCounterOnTap: Bool = false
     ) {
+        let onTap = showCounterOnTap ? { [weak self] in
+            self?.present(CounterViewController(), animated: true)
+        } : nil
+        
```
At first, I tried this instead:
```
let onTap = showCounterOnTap ? { [weak self] in self?.showCounterView() } : nil
```
And I kept the function on lines 125-128 (I did, however, remove `@objc` since it's no longer needed).

However, this still led to a memory leak. Is this because there is still a strong reference to `self.present()` within the body of `showCounterView()`?
